Add type hints

AttackingOrDefending opened the pull request:
Adds type hints to all functions.

AttackingOrDefending commented:
This should be ready for review.
I only added type hints to all lines with `def`, not to every variable. This is intended for another PR.
I didn't add mypy to github actions because to get the number of errors down to 0, will require a lot of changes to the existing code.
There are 191 mypy errors remaining (using `mypy . --strict --exclude=.venv`).
In `lichess-bot.py`, all the different `Dict[str, Any]`s were added, so that the user will be able to differentiate between the different dicts. For mypy they are the exact same.
I used `queue.Queue` and `list` event though we use a proxy of them, because I couldn't use the proxies, but they work the exact same way.
To use `dict`, `list`, `x | None`, `x | y` instead of `Dict`, `List`, `Optional[x]`, `Union[x, y]`, would require dropping support for python 3.7 and 3.8.

MarkZH commented on config.py:
```diff
@@ -3,41 +3,50 @@
 logger = logging.getLogger(__name__)
 
 
+class DelayType(str, Enum):
```
Why was `DelayType` copied here?

AttackingOrDefending replied:
Because there was a circular import. This is why `MAX_CHAT_MESSAGE_LEN` was also copied. Because I added `from config import Configuration` in matchmaking.py I couldn't import anything from matchmaking.

MarkZH replied:
For `MAX_CHAT_MESSAGE_LEN`, you can delete the definition in engine_wrapper.py and use `lichess.MAX_CHAT_MESSAGE_LEN` inside the `get_stats()` method.

Similarly for `DelayType`, delete the definition in matchmaking.py and use `config.DelayType` instead.

This removes duplicate definitions.

AttackingOrDefending commented:
The new tests for python 3.8 found some errors here. They were easy to fix, but unfortunately the quotes will have to be kept in the types initialized in `lichess-bot.py` (e.g. `CONTROL_QUEUE_TYPE = "queue.Queue[EVENT_TYPE]"`), because python 3.8 doesn't support using `Queue[...]`.

MarkZH commented:
Is the idea here like the flake8: put the structure in place and then slowly fix up the errors revealed?

AttackingOrDefending commented:
I added mypy on github actions, but it will always be failing until we fix all the erros. It is somewhat like `flake8` (as you can also see [here](https://mypy.readthedocs.io/en/stable/existing_code.html)), but it is hard to fix all the errors. Also, `lichess-bot.py` will always raise an error, because I added quotes around some types as mentioned in my previous comment, until we drop support for `python 3.8`.
